Create Reusable Sidebar Navigator Components for Calypso

yashwin opened the pull request:
Resolves https://github.com/Automattic/jetpack-genesis/issues/57

## Proposed Changes

This PR adds Navigator components(SidebarNavigator, SidebarNavigatorMenu & SidebarNavigatorMenuItem) to the existing SidebarV2 component. 

https://github.com/Automattic/wp-calypso/assets/10586875/5086c2f1-c57d-477b-873e-a96f8e1c8b3e

### Testing Instructions

**Prerequisites**

Since these changes are made specifically for agencies, you must set yourself(partner) as an agency - 2c49b-pb. Make sure to switch it back to the previous type.

**Instructions**

- Switch to the `add/implement-sidebar-navigation` branch > Set `jetpack/new-navigation` to true in `config/jetpack-cloud-development.json` file >  Start the server 
- Go to `client/jetpack-cloud/sections/sidebar-navigation/jetpack-manage.tsx` and make the below changes:

Feel free to add more items if needed.

```
import { key, currencyDollar } from '@wordpress/icons';
import { useTranslate } from 'i18n-calypso';
import page from 'page';
import NewSidebar from 'calypso/jetpack-cloud/components/sidebar';

const JetpackManageSidebar = () => {
	const translate = useTranslate();

	const onClickMenuItem = ( path: string ) => {
		page.redirect( path );
	};

	const menuItems = [
		{
			icon: key,
			path: '/',
			link: '/partner-portal/licenses',
			title: translate( 'Licenses' ),
			onClickMenuItem: onClickMenuItem,
                         isSelected: true,
		},
		{
			icon: currencyDollar,
			path: '/partner-portal/',
			link: '/partner-portal/billing',
			title: translate( 'Purchases' ),
			onClickMenuItem: onClickMenuItem,
			withChevron: true,
		},
	];
	return <NewSidebar isJetpackManage path="/" menuItems={ menuItems } />;
};

export default JetpackManageSidebar;
```

- Go to `client/jetpack-cloud/sections/sidebar-navigation/purchases.tsx` and make the below changes

```
import { store, chevronLeft } from '@wordpress/icons';
import { useTranslate } from 'i18n-calypso';
import page from 'page';
import NewSidebar from 'calypso/jetpack-cloud/components/sidebar';

const PurchasesSidebar = () => {
	const translate = useTranslate();

	const onClickMenuItem = ( path: string ) => {
		page.redirect( path );
	};

	const menuItems = [
		{
			icon: store,
			path: '/partner-portal',
			link: '/partner-portal/billing',
			title: translate( 'Billing' ),
			onClickMenuItem: onClickMenuItem,
                          isSelected: true
		},
	];
	return (
		<NewSidebar
			isJetpackManage
			path="/partner-portal"
			menuItems={ menuItems }
			description={ translate( 'Manage all your billing related settings from one place.' ) }
			backButtonProps={ {
				label: translate( 'Purchases' ),
				icon: chevronLeft,
				onClick: () => onClickMenuItem( '/dashboard' ),
			} }
		/>
	);
};

export default PurchasesSidebar;
```

- Verify that you can see the sidebar menu items loaded for /dashboard

<img width="271" alt="Screenshot 2023-10-17 at 11 25 08 AM" src="https://github.com/Automattic/wp-calypso/assets/10586875/9c5caced-e863-4290-ae9e-c525cf6a20c2">


- Click on the Purchases tab > Verify that the billing page is loaded and you can see the updated sidebar menu items for the Purchases tab along with back button and the description.

<img width="270" alt="Screenshot 2023-10-17 at 11 24 44 AM" src="https://github.com/Automattic/wp-calypso/assets/10586875/0d15df6e-ad0a-4925-a56d-c6fd30731577">

- Now, refresh the page and verify that the current page is highlighted and the sidebar items are shown as expected
- Click the back button (`< Purchases`) and verify that you are redirected to the Dashboard and the sidebar menu items are updated accordingly. 

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

yashwin committed:
Create SidebarNavigator component

yashwin committed:
Create SidebarNavigatorMenu component

yashwin committed:
Create SidebarNavigatorMenuItem component

yashwin committed:
Export other navigations components from the index.tsx file

yashwin committed:
Export navigator components from the "/sidebar-v2/index.tsx" file

yashwin committed:
Use the navigator components to render the sidebar in JetpackCloudSidebar

yashwin committed:
Add the required props to NewSidebar component

yashwin commented on .eslintrc.js:
```diff
@@ -499,6 +499,7 @@ module.exports = {
 					'__experimentalItemGroup',
 					'__experimentalNavigationBackButton',
 					'__experimentalNavigatorBackButton',
+					'__experimentalNavigatorToParentButton',
```
Since this is an experimental API, we need to disable the linter warning in the `.eslintrc.js` file so that we don’t see the `wpcalypso/no-unsafe-wp-apis` warning.

jkguidaven commented on client/layout/sidebar-v2/navigator/style.scss:
```diff
@@ -0,0 +1,37 @@
+	&.is-active {
+		border-radius: 4px !important;
+		background: var(--studio-green-0) !important;
+		color: var(--studio-green-50) !important;
```
For Jetpack-specific styling, we should override this from the consuming component instead of the base component.

yashwin replied:
Makes sense. Moved it here: https://github.com/Automattic/wp-calypso/pull/83073/commits/e655f6f14d2a89da4708c2f088a6af10a76b9f59

jkguidaven commented on client/layout/sidebar-v2/navigator/navigator-menu-item.tsx:
```diff
@@ -0,0 +1,55 @@
+				<HStack justify="flex-start">
+					{ icon && <Icon style={ { fill: 'currentcolor' } } icon={ icon } size={ ICON_SIZE } /> }
+					<FlexBlock>{ children }</FlexBlock>
+					{ withChevron && <Icon icon={ chevronRightSmall } size={ ICON_SIZE } /> }
```
Can we also add an optional right icon or `isExternalLink` to show an external icon like the use case below?

<img width="495" alt="Screen Shot 2023-10-17 at 3 13 58 PM" src="https://github.com/Automattic/wp-calypso/assets/56598660/cc85775e-ec15-49eb-90a1-06fc8e762555">

yashwin replied:
Thanks for the suggestion. Added it here: https://github.com/Automattic/wp-calypso/pull/83073/commits/9b5d6043e59f010e9a2f9bdfcc54da0ed61ef889

jkguidaven replied:
Thanks for addressing this use case. 🙇

jkguidaven commented:
It seems that the sidebar and sidebar menu backgrounds are not exactly the same shade of white.


<img src="https://user-images.githubusercontent.com/10586875/275726058-0d15df6e-ad0a-4925-a56d-c6fd30731577.png" />

yashwin committed:
Move Jetpack Cloud specific styles

yashwin committed:
Add optional external icon

yashwin committed:
Update navigator background color

yashwin commented:
@jkguidaven: Thanks for the review!

I have addressed all your comments.

> It seems that the sidebar and sidebar menu backgrounds are not exactly the same shade of white.

I initially thought some parent component was doing it. Fixed it here: https://github.com/Automattic/wp-calypso/pull/83073/commits/bec4534140d5c22b2b45d6e8f219458dade58c3b

sejas commented on client/layout/sidebar-v2/navigator/style.scss:
```diff
@@ -0,0 +1,45 @@
+	padding: 0 !important;
+
+	.components-navigator-back-button {
+		margin-block-end: 8px !important;
+		color: var(--studio-gray-90) !important;
```
Would be possible to remove these `!important` from the CSS?

mattgawarecki replied:
Big +1 to this -- `!important` is heavily discouraged across Calypso. If necessary, one alternative approach is to repeat the same class name in the rule specification, like so:

```css
/* Upstream rule we'd like to override */
/* Specificity is 0-2-0 */
.some-parent .with-a-child-my-class {
	font-size: normal;
}
```

```css
/* Downstream rule we'd like to use */
/* Before: specificity is 0-1-0, fails to override */
.my-class {
	font-size: large;
}
```

```css
/* Downsteram rule we'd like to use */
/* After: specificity is now 0-3-0, override successful */
.my-class.my-class.my-class {
	font-size: large;
}
```

yashwin replied:
I totally agree with this. I think I missed it earlier. I fixed it now: https://github.com/Automattic/wp-calypso/pull/83073/commits/64e64879cad9a18cd1fa3bdec209bbd2746d7803. Thanks!

sejas reviewed (commented):
@yashwin , great work!

I tested it and it works as described.

https://github.com/Automattic/wp-calypso/assets/779993/079ea3d7-57c0-4f78-8c8d-1e6d632baa53

Will be possible to add the animation that appears in the design?

p1HpG7-nGl-p2#comment-66162 

https://github.com/Automattic/wp-calypso/assets/779993/495200e6-6239-43c0-9bda-5fb404e7718a




I wonder if our menus should implement  `SidebarNavigatorMenu` directly, where all of them share the same `SidebarNavigator`.
I'm using this storybook as reference : https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-navigator--docs

Something like this:

```tsx
import { store, chevronLeft, page as pageIcon } from '@wordpress/icons';
import { useTranslate } from 'i18n-calypso';
import page from 'page';
import { SidebarNavigatorMenu, SidebarNavigatorMenuItem } from 'calypso/layout/sidebar-v2';

const PurchasesSidebar = () => {
	const translate = useTranslate();

	const onClickMenuItem = ( path: string ) => {
		page.redirect( path );
	};

	const menuItems = [
		{
			icon: store,
			path: '/partner-portal',
			link: '/partner-portal/billing',
			title: translate( 'Billing' ),
			onClickMenuItem: onClickMenuItem,
		},
		{
			icon: pageIcon,
			path: '/partner-portal',
			link: '/partner-portal/invoices',
			title: translate( 'Invoices' ),
			onClickMenuItem: onClickMenuItem,
		},
	];
	return (
		<SidebarNavigatorMenu
			path="/partner-portal"
			description={ translate( 'Manage all your billing related settings from one place.' ) }
			backButtonProps={ {
				label: translate( 'Purchases' ),
				icon: chevronLeft,
				onClick: () => onClickMenuItem( '/dashboard' ),
			} }
		>
			{ menuItems.map( ( item ) => (
				<SidebarNavigatorMenuItem key={ item.link } { ...item } />
			) ) }
		</SidebarNavigatorMenu>
	);
};

export default PurchasesSidebar;

// In another file
<SidebarNavigator initialPath={ path }>
  <RootMenuWithLicenseAndPurchases />
  <PurchasesSidebar />
</DSidebarNavigator>


```


Instead of the suggested in the PR description:

```tsx
import { store, chevronLeft, page as pageIcon } from '@wordpress/icons';
import { useTranslate } from 'i18n-calypso';
import page from 'page';
import NewSidebar from 'calypso/jetpack-cloud/components/sidebar';

const PurchasesSidebar = () => {
	const translate = useTranslate();

	const onClickMenuItem = ( path: string ) => {
		page.redirect( path );
	};

	const menuItems = [
		{
			icon: store,
			path: '/partner-portal',
			link: '/partner-portal/billing',
			title: translate( 'Billing' ),
			onClickMenuItem: onClickMenuItem,
		},
		{
			icon: pageIcon,
			path: '/partner-portal',
			link: '/partner-portal/invoices',
			title: translate( 'Invoices' ),
			onClickMenuItem: onClickMenuItem,
		},
	];
	return (
		<NewSidebar
			isJetpackManage
			path="/partner-portal"
			menuItems={ menuItems }
			description={ translate( 'Manage all your billing related settings from one place.' ) }
			backButtonProps={ {
				label: translate( 'Purchases' ),
				icon: chevronLeft,
				onClick: () => onClickMenuItem( '/dashboard' ),
			} }
		/>
	);
};

export default PurchasesSidebar;
```

yashwin committed:
Do not use !important

yashwin commented:
@sejas: Thanks for the review! I have addressed your comments and made the changes where needed.

> Will be possible to add the animation that appears in the design?

Good point. I think this can be done in another PR as it needs some R&D IMO, so I have created a task for the same: https://github.com/Automattic/jetpack-genesis/issues/66, which we will address before shipping this project.

> I wonder if our menus should implement SidebarNavigatorMenu directly, where all of them share the same SidebarNavigator.

I prefer the existing implementation. I'm not sure how complicated it is to change how the sidebar and the layout are rendered currently.

yashwin committed:
Use isSelected prop to show the active menu item

yashwin committed:
Create README.md

yashwin commented:
@jkguidaven @sejas @mattgawarecki: I have made a few changes:

- We are now accepting the `isSelected` prop for SidebarNavigatorMenuItem since we have a lot of dynamic URL  https://github.com/Automattic/wp-calypso/pull/83073/commits/e5c8c9dfc1937ad8aa2173ff9241a3cf097c0a42
- I have created a README.md: https://github.com/Automattic/wp-calypso/pull/83073/commits/f902186c5c04673e657613a96da9bd082033d464

sejas reviewed (approved):
> @sejas: Thanks for the review! I have addressed your comments and made the changes where needed.
> 
> > Will be possible to add the animation that appears in the design?
> 
> Good point. I think this can be done in another PR as it needs some R&D IMO, so I have created a task for the same: https://github.com/Automattic/jetpack-genesis/issues/66, which we will address before shipping this project.
> 
> > I wonder if our menus should implement SidebarNavigatorMenu directly, where all of them share the same SidebarNavigator.
> 
> I prefer the existing implementation. I'm not sure how complicated it is to change how the sidebar and the layout are rendered currently(the way the sidebar is set to the context). If you are curious and want to implement it like that in WPCOM, I don't think that should be a problem as this is only done for Jetpack Cloud, and these components can be used as per your requirement 🙂

I agree that using a unified navigator for the sidebar will be a bigger refactor, but it will be  probably needed if you want to use the default animation. 

Thanks for creating the shared components flexible so the teams can decide on the final solution ❤️ .

jkguidaven reviewed (approved):
Thanks for addressing my PRs and working. I have nothing else. code LGTM 👍
